pkg/kubelet/pleg/evented.go

```diff
@@ -0,0 +1,291 @@
+/*
+Copyright 2015 The Kubernetes Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package pleg
+
+import (
+	"fmt"
+
+	"k8s.io/apimachinery/pkg/types"
+	"k8s.io/apimachinery/pkg/util/wait"
+	internalapi "k8s.io/cri-api/pkg/apis"
+	runtimeapi "k8s.io/cri-api/pkg/apis/runtime/v1"
+	"k8s.io/klog/v2"
+	kubecontainer "k8s.io/kubernetes/pkg/kubelet/container"
+	"k8s.io/kubernetes/pkg/kubelet/metrics"
+	"k8s.io/utils/clock"
+)
+
+type EventedPLEG struct {
+	// The container runtime.
+	runtime kubecontainer.Runtime
+
+	runtimeService internalapi.RuntimeService
+
+	// The channel from which the subscriber listens events.
+	eventChannel chan *PodLifecycleEvent
+
+	// Cache for storing the runtime states required for syncing pods.
+	cache kubecontainer.Cache
+	// For testability.
+	clock clock.Clock
+}
+
+// NewEventedPLEG instantiates a new EventedPLEG object and return it.
+func NewEventedPLEG(runtime kubecontainer.Runtime, runtimeService internalapi.RuntimeService, eventChannel chan *PodLifecycleEvent,
+	cache kubecontainer.Cache, clock clock.Clock) PodLifecycleEventGenerator {
+	return &EventedPLEG{
+		runtime:        runtime,
+		runtimeService: runtimeService,
+		eventChannel:   eventChannel,
+		cache:          cache,
+		clock:          clock,
+	}
+}
+
+// Watch returns a channel from which the subscriber can receive PodLifecycleEvent events.
+// TODO: support multiple subscribers.
+func (e *EventedPLEG) Watch() chan *PodLifecycleEvent {
+	return e.eventChannel
+}
+
+// Start spawns a goroutine to relist periodically.
+func (e *EventedPLEG) Start() {
+	go wait.Forever(e.watchEventsChannel, 0)
+}
+
+// Healthy check if PLEG work properly.
+func (e *EventedPLEG) Healthy() (bool, error) {
+	// GenericPLEG is declared unhealthy when relisting time is more than
+	// the relistThreshold. In case, EventedPLEG is turned on the value of
+	// relistingPeriod, relistingThreshold is adjusted to higher values and
+	// so health of Generic PLEG should take care of the condition checking
+	// higher relistThreshold between last two relist.
+
+	// EventedPLEG is declared unhealthy only if eventChannel is out of capacity.
```

Comment:
i would remove the channel check until we better understand how quickly this channel can hit 1k entries.

can we remove this for the moment until we gather metrics on channel size to avoid a node going not ready?  i am trying to think about a node reboot (with no drain of pods) and the number of concurrent events that could happen at that time.

Reply:
`Healthy()` function has to be implemented by any object implementing  `PodLifecycleEventGenerator` interface
https://github.com/kubernetes/kubernetes/pull/111384/files#diff-a33847e3f84b1d1dc42c7ed04ad9fc3d7ace8fafd4aa5670b31ddc5f22e5345cR56

Although in case of `EventedPLEG` this `Healthy()` function is not called by anyone yet.